test/many-streams.test.js

```diff
@@ -131,40 +131,39 @@ async function makeTableP({gbqClient, gbqDatasetName, gbqTableName}) {
 }
 
 function createReadStream() {
+  var results = Array(streamLength).fill(stringifyObject(testData))
 
   return new Readable({
-
-    objectMode   : true,
-    highWaterMark: streamLength,
-
     read(_size) {
-
       const thisStream = this
+      this.push(results.pop())
 
-      ramda.range(0, streamLength).map(() => {
-
-        let obj = testData
-
-        if (randomizeData) {
-            obj = ramda.clone(testData)
-            randomizeM(obj)
-        }
-
-        const str = stringifyObject(obj)
-
-        thisStream.push(str)
-      })
-      thisStream.push(null)
+      if (!results.length) {
+        this.push(null)
+      }
     },
   })
 }
 
+var allOutputRows = 0
 function streamsToGcloudPromiseP(readStream, writeStream) {
 
   return new Promise((resolve, reject) => {
 
     writeStream.on('error', reject)
-    writeStream.on('complete', resolve)
+    writeStream.on('complete', job => {
+      job.on('error', reject)
+      job.on('complete', job => {
```

Comment:
This isn't necessary, but using the `job`'s events, we can skip the optimistic delay later in the test, and instead know for certain that the rows were inserted.

This also gives us what turned out to be the key component of this issue, the logging of the `job` object's state. Most of the time, everything works as expected. However, an occasional test run returns:

```
  Rejected promise returned by test. Reason:

  ErrorClass {
    code: 503,
    errors: [
      {
        domain: 'global',
        message: 'Error encountered during execution. Retrying may solve the problem.',
        reason: 'backendError',
      },
    ],
    message: 'Error encountered during execution. Retrying may solve the problem.',
    response: undefined,
  }

  Object.parseHttpRespBody (node_modules/@google-cloud/common/src/util.js:192:30)
  Object.handleResp (node_modules/@google-cloud/common/src/util.js:132:18)
  Request._callback (node_modules/@google-cloud/common/src/util.js:255:14)
  Request.self.callback (node_modules/request/request.js:186:22)
  Request.<anonymous> (node_modules/request/request.js:1163:10)
  IncomingMessage.<anonymous> (node_modules/request/request.js:1085:12)
  Test.__dirname [as fn] (test/many-streams.test.js:189:38)
```

I'm going to take this finding back to the original issue and discuss what we should do about it.

Reply:
Perfect, thank you. I didn't quite understand how to use `job` from the documentation but this makes much more sense.